settings.org
#+TITLE: Jonix Emacs configuration
#+OPTIONS: toc:4 h:4
#+STARTUP: overview

* Core
** Restart Server/Daemon
#+BEGIN_SRC emacs-lisp
  ;; define function to shutdown emacs server instance

  (defun jx/server-stop ()
  (interactive)
  (if (daemonp)
      (save-buffers-kill-emacs)
      (save-buffers-kill-terminal)))
#+END_SRC

** Add personal module
#+BEGIN_SRC emacs-lisp
(add-to-list 'load-path "~/.emacs.d/extra-plugins")
#+END_SRC

** Setup Cask Package Manager
#+BEGIN_SRC emacs-lisp
(require 'cask "~/.cask/cask.el")
(cask-initialize)
;; # Update package depandancy list in ~/.emacs.d/Cask file
;; # To manually install dependencies
;; # Run from .emacs.d folder cask install
;; # (To setup cask run cask init from .emacs.d directory)
#+END_SRC

** Setup use-package for easier Emacs configuration
#+BEGIN_SRC emacs-lisp
(eval-when-compile
  (require 'use-package))
#+END_SRC

** Load better defaults
#+BEGIN_SRC emacs-lisp
  (use-package better-defaults)
#+END_SRC

** Save all buffers when Emacs loses focus
#+BEGIN_SRC emacs-lisp
  (add-hook 'focus-out-hook (lambda () (save-some-buffers t)))

;; automatically save buffers associated with files on buffer switch
;; and on windows switch
;; (defadvice switch-to-buffer (before save-buffer-now activate)
;;   (when buffer-file-name (save-buffer)))
;; (defadvice other-window (before other-window-now activate)
;;   (when buffer-file-name (save-buffer)))
;; (defadvice windmove-up (before other-window-now activate)
;;   (when buffer-file-name (save-buffer)))
;; (defadvice windmove-down (before other-window-now activate)
;;   (when buffer-file-name (save-buffer)))
;; (defadvice windmove-left (before other-window-now activate)
;;   (when buffer-file-name (save-buffer)))
;; (defadvice windmove-right (before other-window-now activate)
;;   (when buffer-file-name (save-buffer)))

#+END_SRC

** Show column number
#+BEGIN_SRC emacs-lisp
(setq column-number-mode t)

;; Force Emacs to show first column as 1 instead of 0
;; Might slow Emacs down, since it uses 'force-mode-line-update'
;;
;; update the mode line to have line number and column number
(setq mode-line-position
      '("%p (%l," (:eval (format "%d)" (1+ (current-column))))))
;; force the update of the mode line so the column gets updated
(add-hook 'post-command-hook 'force-mode-line-update)
#+END_SRC

** Bookmark
#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "C-x r j") 'bookmark-jump)
#+END_SRC

** Enable recent file menu
#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "C-c r") 'recentf-open-files)
#+END_SRC

** Make scripts executable
#+BEGIN_SRC emacs-lisp
(add-hook 'after-save-hook
  'executable-make-buffer-file-executable-if-script-p)
#+END_SRC

** Browse kill ring
#+BEGIN_SRC emacs-lisp
  (use-package browse-kill-ring
    :bind ("C-c k" . browse-kill-ring))
#+END_SRC

** Inhibit start-screen
#+BEGIN_SRC emacs-lisp
(setq inhibit-startup-message t)
(setq inhibit-splash-screen t)
(setq inhibit-startup-echo-area-message t)
#+END_SRC

** Load theme
#+BEGIN_SRC emacs-lisp
; (load-theme 'solarized-dark t)
(load-theme 'zenburn t)
#+END_SRC

** Turn off menu bar
#+BEGIN_SRC emacs-lisp
(menu-bar-mode -1)
#+END_SRC

** Empty initial scratch buffer
#+BEGIN_SRC emacs-lisp
(setq initial-scratch-message "")
#+END_SRC

** Cycling start of line to either indentation level or start of line
#+BEGIN_SRC emacs-lisp
(defun mak::cycle-bol-boi ()
  "Cycle between beginning of line and beginning of indentation."
  (interactive)
  (let ((orig (point)))
    (back-to-indentation)
    (when (= orig (point))
      (move-beginning-of-line 1))))

(global-set-key (kbd "C-a") 'mak::cycle-bol-boi)
(global-set-key (kbd "<home>") 'mak::cycle-bol-boi)
#+END_SRC
** Company
#+BEGIN_SRC emacs-lisp
(add-hook 'after-init-hook 'global-company-mode)
;; To see or change the list of enabled back-ends, type M-x customize-variable RET company-backends. Also see its description for information on writing a back-end.
;; For more information, type M-x describe-function RET company-mode.

;; 2016-07-27 - Trying to suppress irony php warning
;(eval-after-load 'company
;    '(add-to-list 'company-backends '(company-irony company-jedi))
;)
(eval-after-load 'company
    '(add-to-list 'company-backends '(company-jedi))
)

#+END_SRC

** Smex
#+BEGIN_SRC emacs-lisp

  (use-package smex
    :ensure t
    :bind (
           ("M-x" . smex)
           ("M-X" . smex-major-ode-commands)
           ("C-c C-c M-x" . execute-extended-command)
          )
    :config (smex-initialize)
   )

#+END_SRC

** Increase font-size
#+BEGIN_SRC emacs-lisp
;; Font size
(global-set-key (kbd "C-+") 'text-scale-increase)
(global-set-key (kbd "C--") 'text-scale-decrease)
#+END_SRC

** Recent file support
#+BEGIN_SRC emacs-lisp
;; Enable recent-file
(require 'recentf)

;; get rid of `find-file-read-only' and replace it with something more useful.
(global-set-key (kbd "C-x C-r") 'ido-recentf-open)

;; enable recent files mode.
(recentf-mode t)

; 64 files ought to be enough for everyone ;-)
(setq recentf-max-saved-items 64)

(defun ido-recentf-open ()
  "Use `ido-completing-read' to \\[find-file] a recent file"
  (interactive)
  (if (find-file (ido-completing-read "Find recent file: " recentf-list))
      (message "Opening file...")
    (message "Aborting")))

#+END_SRC

** Disable suspend-emacs
#+BEGIN_SRC emacs-lisp
;; Disable suspend-emacs
(global-set-key "\C-z" nil)
(global-set-key "\C-x\C-z" nil)
#+END_SRC

** Disable Exit emacs
** Remove added whitespace
#+BEGIN_SRC emacs-lisp
(setq whitespace-cleanup-mode t)
#+END_SRC

** Easy undo window configuration
#+BEGIN_SRC emacs-lisp
;; Enable to store window configuration
(winner-mode 1)
#+END_SRC

** God mode
#+BEGIN_SRC emacs-lisp
;; Enable God mode (avoid pressing Ctrl key all the time)
;; Enters a special mode
(require 'god-mode)

(defun my-update-cursor ()
  (setq cursor-type (if (or god-local-mode buffer-read-only)
                      'bar
                      'box)))
(add-hook 'god-mode-enabled-hook 'my-update-cursor)
(add-hook 'god-mode-disabled-hook 'my-update-cursor)

;; Make a tweak to split windows using God-mode
(global-set-key (kbd "C-x C-1") 'delete-other-windows)
(global-set-key (kbd "C-x C-2") 'split-window-below)
(global-set-key (kbd "C-x C-3") 'split-window-right)
(global-set-key (kbd "C-x C-0") 'delete-window)
;; Enable repeat
(define-key god-local-mode-map (kbd ".") 'repeat)

;; Enable God-mode on isearch
;(require 'god-mode-isearch)
;(define-key isearch-mode-map (kbd "<escape>") 'god-mode-isearch-activate)
;(define-key god-mode-isearch-map (kbd "<escape>") 'god-mode-isearch-disable)

;; Bind Caps-lock to M-x under GNU/Linux
;; From http://emacs-fu.blogspot.se/2008/12/remapping-caps-lock.html
(if (eq system-type 'gnu/linux)
    ;(shell-command "xmodmap -e 'clear Lock' -e 'keycode 66 = F13'")
  ;; Requires the bash command "xmodmap -e 'clear Lock' -e 'keycode 66 = F13'" to be run prior to this binding
  (global-set-key [f13] 'god-mode-all)
)

;; Bind Caps-lock to M-x For Windows
(if (eq system-type 'windows-nt)
    (setq w32-enable-caps-lock nil)
    ;(global-set-key [capslock] 'god-local-mode)
    (global-set-key [capslock] 'god-mode-all)
)

#+END_SRC

** Date and time
*** Add week number to Emacs calendar view
#+BEGIN_SRC emacs-lisp
;;; Calender stuff

;; Add week number to Emacs calender
(copy-face font-lock-constant-face 'calendar-iso-week-face)
(set-face-attribute 'calendar-iso-week-face nil
                    :height 0.7)
(setq calendar-intermonth-text
      '(propertize
        (format "%2d"
                (car
                 (calendar-iso-from-absolute
                  (calendar-absolute-from-gregorian (list month day year)))))
        'font-lock-face 'calendar-iso-week-face))

;; End of calender stuff
#+END_SRC

*** Insert Date string
#+BEGIN_SRC emacs-lisp
;; Insert date at current position
(defun jx/current-date () (interactive)
    (insert (shell-command-to-string "echo -n $(date +%Y-%m-%d)")))

;; Insert time at current position
(defun jx/current-time () (interactive)
 (insert (shell-command-to-string "echo -n $(date +%H:%M)")))

#+END_SRC

*** Insert Week nr
#+BEGIN_SRC emacs-lisp
;; Insert week number at current position
(defun jx/week-nr () (interactive)
    (insert (shell-command-to-string "echo -n $(date +%W)")))
#+END_SRC
*** Insert name of day
#+BEGIN_SRC emacs-lisp
;; Insert day of name at current position
(defun jx/day () (interactive)
    (insert (shell-command-to-string "echo -n $(date +%A)")))
#+END_SRC


*** Show clock
#+BEGIN_SRC emacs-lisp
;; Show clock in status-bar
(setq display-time t
      display-time-24hr-format t)
(display-time)
#+END_SRC

** Keybindings help
#+BEGIN_SRC emacs-lisp
(which-key-mode)
#+END_SRC

** Scrolling margin
When the cursor is on the top or bottom of the screen,
and it is time to scroll, display x lines of context
#+BEGIN_SRC emacs-lisp
; (setq scroll-margin 0)
#+END_SRC

** Dired settings
*** Make dired less verbose
- Cask kunde inte ladda ned detta paket 2021-07-25
  Disablar paketet
# #+BEGIN_SRC emacs-lisp
# ;; Make dired less verbose
# (require 'dired-details)
# (setq-default dired-details-hidden-string "--- ")
# (dired-details-install)
# #+END_SRC

*** Dired Reuse buffers
http://ergoemacs.org/emacs/emacs_dired_tips.html
#+BEGIN_SRC emacs-lisp
(require 'dired-x)
(setq dired-omit-files "^\\.?#\\|^\\.$\\|^\\.\\.$\\|^\\.")
(add-hook 'dired-mode-hook (lambda ()
                             (dired-omit-mode 1)))
(setq dired-listing-switches "-aBhl --group-directories-first")
(put 'dired-find-alternate-file 'disabled nil)

; Let <enter> re-use dired directory buffer
(define-key dired-mode-map (kbd "RET") 'dired-find-alternate-file) ; was dired-advertised-find-file

;; Let <u> (go up one directroy) re-use directory buffer when
(add-hook 'dired-mode-hook
 (lambda ()
  (define-key dired-mode-map (kbd "u")
    (lambda () (interactive) (find-alternate-file "..")))))

;; Let <backspace> (go up one directroy) re-use directory buffer when
(add-hook 'dired-mode-hook
 (lambda ()
  (define-key dired-mode-map (kbd "<DEL>")
    (lambda () (interactive) (find-alternate-file "..")))))

(eval-after-load "dired"
  '(progn
     (defadvice dired-advertised-find-file (around dired-subst-directory activate)
       "Replace current buffer if file is a directory."
       (interactive)
       (let* ((orig (current-buffer))
              ;; (filename (dired-get-filename))
              (filename (dired-get-filename t t))
              (bye-p (file-directory-p filename)))
         ad-do-it
         (when (and bye-p (not (string-match "[/\\\\]\\.$" filename)))
           (kill-buffer orig))))))

#+END_SRC
*** Filter Dired files
Filter on files in Dired mode
Press / to start filtering
Press g to revert
#+BEGIN_SRC emacs-lisp
  (eval-after-load
      "dired" '(progn
  (define-key dired-mode-map (kbd "/") 'dired-narrow)))
#+END_SRC

*** Toggle hidden files view in Dired mode
Press . to toggle view of hidden files
#+BEGIN_SRC emacs-lisp
  (defun dired-toggle-dotfiles ()
    "Show/hide dot-files"
    (interactive)
    (when (equal major-mode 'dired-mode)
      (if (or (not (boundp 'dired-dotfiles-show-p)) dired-dotfiles-show-p) ; if currently showing
          (progn
            (set (make-local-variable 'dired-dotfiles-show-p) nil)
            (message "h")
            (dired-mark-files-regexp "^\\\.")
            (dired-do-kill-lines))
        (progn (revert-buffer) ; otherwise just revert to re-show
               (set (make-local-variable 'dired-dotfiles-show-p) t)))))


(eval-after-load "dired" '(progn
  (define-key dired-mode-map (kbd ".") 'dired-toggle-dotfiles)))

 (global-set-key [f7] 'neotree-project-dir)

#+END_SRC

*** Jump to current file in Dired
In any file buffer, call dired-jump 【Ctrl+x Ctrl+j】 to jump to the directory of current buffer.
#+BEGIN_SRC emacs-lisp
(require 'dired-x)
#+END_SRC

*** Copy / Delete directories
#+BEGIN_SRC emacs-lisp
;; allow dired to delete or copy dir
(setq dired-recursive-copies (quote always)) ; “always” means no asking
(setq dired-recursive-deletes (quote top)) ; “top” means ask once
#+END_SRC

*** Move files between split panes
#+BEGIN_SRC emacs-lisp
(setq dired-dwim-target t)
#+END_SRC
*** Reload dired after making changes
#+BEGIN_SRC emacs-lisp
(--each '(dired-do-rename
          dired-do-copy
          dired-create-directory
          wdired-abort-changes)
        (eval `(defadvice ,it (after revert-buffer activate)
(revert-buffer))))
#+END_SRC
** Go to last change in buffer
 - 'C-c b ,' Go to last change
 - 'C-c b .' Go to previous change

#+BEGIN_SRC emacs-lisp
(require 'goto-chg)
(global-set-key (kbd "C-c b ,") 'goto-last-change)
(global-set-key (kbd "C-c b .") 'goto-last-change-reverse)
#+END_SRC

** Edit current file with Sudo privs
#+BEGIN_SRC emacs-lisp
(require 'sudo-edit)
#+END_SRC

** Discover key-bindings for major/minor modes
#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "C-h C-m") 'discover-my-major)
(global-set-key (kbd "C-h M-m") 'discover-my-mode)
#+END_SRC

** Easy kill
easy-kill is a drop-in replacement for kill-ring-save.
Included is easy-mark

Keys:
    M-w w: save word at point
    M-w s: save sexp at point
    M-w l: save list at point (enclosing sexp)
    M-w d: save defun at point
    M-w D: save current defun name
    M-w f: save file at point
    M-w b: save buffer-file-name or default-directory. - changes the kill to the directory name, + to full name and 0 to basename.

The following keys modify the selection:

    @: append selection to previous kill and exit. For example, M-w d @ will append current function to last kill.
    C-w: kill selection and exit
    +, - and 1..9: expand/shrink selection
    0 shrink the selection to the initial size i.e. before any expansion
    C-SPC: turn selection into an active region
    C-g: abort
    ?: help

#+BEGIN_SRC emacs-lisp
(global-set-key [remap kill-ring-save] 'easy-kill)
#+END_SRC

** Cleanup whitespace on save
#+BEGIN_SRC emacs-lisp
(add-hook 'before-save-hook 'whitespace-cleanup)
#+END_SRC

** Expand region
#+BEGIN_SRC emacs-lisp
(require 'expand-region)
(global-set-key (kbd "C-x =") 'er/expand-region)
#+END_SRC

** Multiple major mode in the same buffer
#+BEGIN_SRC emacs-lisp
;; Using the package polymode
(require 'poly-markdown)
(require 'poly-org)
(add-to-list 'auto-mode-alist '("\\.md$" . poly-markdown-mode))
(add-to-list 'auto-mode-alist '("\\.org$" . poly-org-mode))
#+END_SRC
** YASnippets
#+BEGIN_SRC emacs-lisp
(require 'yasnippet)
(add-to-list 'yas-snippet-dirs "~/yas-snippets")
(yas-global-mode t)



#+END_SRC

** Load very large file
#+BEGIN_SRC emacs-lisp
(require 'vlf-setup)
#+END_SRC

** Find files in project
#+BEGIN_SRC emacs-lisp
(autoload 'find-file-in-project "find-file-in-project" nil t)
(autoload 'find-file-in-project-by-selected "find-file-in-project" nil t)
(autoload 'find-directory-in-project-by-selected "find-file-in-project" nil t)
(autoload 'ffip-show-diff "find-file-in-project" nil t)
(autoload 'ffip-save-ivy-last "find-file-in-project" nil t)
(autoload 'ffip-ivy-resume "find-file-in-project" nil t)
;(defun maybe-project-find-file ()
;  (interactive)
;  (call-interactively
;   (if (projectile-project-p)
;       #'find-file-in-project-by-selected
;       #'ido-find-file)))

(global-set-key (kbd "\C-co") 'find-file-in-project)
#+END_SRC

** Rename file and buffer
#+BEGIN_SRC emacs-lisp
(defun jx/rename-file-and-buffer ()
  "Rename the current buffer and file it is visiting."
  (interactive)
  (let ((filename (buffer-file-name)))
    (if (not (and filename (file-exists-p filename)))
        (message "Buffer is not visiting a file!")
      (let ((new-name (read-file-name "New name: " filename)))
        (cond
         ((vc-backend filename) (vc-rename-file filename new-name))
         (t
          (rename-file filename new-name t)
          (set-visited-file-name new-name t t)))))))

#+END_SRC

** Move file in filesystem
#+BEGIN_SRC emacs-lisp
(defun jnx/move-file (new-location)
  "Write this file to NEW-LOCATION, and delete the old one."
  (interactive (list (if buffer-file-name
                       (read-file-name "Move file to: ")
                       (read-file-name "Move file to: "
                                       default-directory
                                       (expand-file-name (file-name-nondirectory (buffer-name))

settings.el

;; Personal modules outside of Melpa
(add-to-list 'load-path "~/.emacs.d/personal")

(add-hook 'after-save-hook
  'executable-make-buffer-file-executable-if-script-p)

(require 'cask "~/.cask/cask.el")
(cask-initialize)
;; # Update package depandancy list in ~/.emacs.d/Cask file
;; # To manually install dependencies
;; # Run from .emacs.d folder cask install
;; # (To setup cask run cask init from .emacs.d directory)

(require 'better-defaults)

; (load-theme 'solarized-dark t)
(load-theme 'zenburn t)

(menu-bar-mode -1)

(require 'smex)
(smex-initialize)

(global-set-key (kbd "M-x") 'smex)
(global-set-key (kbd "M-X") 'smex-major-mode-commands)
; This is your old M-x.
(global-set-key (kbd "C-c C-c M-x") 'execute-extended-command)

;; Font size
(global-set-key (kbd "C-+") 'text-scale-increase)
(global-set-key (kbd "C--") 'text-scale-decrease)

;; Enable recent-file
(require 'recentf)

;; get rid of `find-file-read-only' and replace it with something more useful.
(global-set-key (kbd "C-x C-r") 'ido-recentf-open)

;; enable recent files mode.
(recentf-mode t)

; 64 files ought to be enough for everyone ;-)
(setq recentf-max-saved-items 64)

(defun ido-recentf-open ()
  "Use `ido-completing-read' to \\[find-file] a recent file"
  (interactive)
  (if (find-file (ido-completing-read "Find recent file: " recentf-list))
      (message "Opening file...")
    (message "Aborting")))

;; Disable Prelude whitespace marking
;; (setq prelude-whitespace nil)

(setq whitespace-cleanup-mode t)

;; Enable to store window configuration
(winner-mode 1)

;; Enable God mode (avoid pressing Ctrl key all the time)
;; Enters a special mode
(require 'god-mode)

(defun my-update-cursor ()
  (setq cursor-type (if (or god-local-mode buffer-read-only)
                      'bar
                      'box)))
(add-hook 'god-mode-enabled-hook 'my-update-cursor)
(add-hook 'god-mode-disabled-hook 'my-update-cursor)

;; Make a tweak to split windows using God-mode
(global-set-key (kbd "C-x C-1") 'delete-other-windows)
(global-set-key (kbd "C-x C-2") 'split-window-below)
(global-set-key (kbd "C-x C-3") 'split-window-right)
(global-set-key (kbd "C-x C-0") 'delete-window)
;; Enable repeat
(define-key god-local-mode-map (kbd ".") 'repeat)

;; Enable God-mode on isearch
;(require 'god-mode-isearch)
;(define-key isearch-mode-map (kbd "<escape>") 'god-mode-isearch-activate)
;(define-key god-mode-isearch-map (kbd "<escape>") 'god-mode-isearch-disable)

;; Bind Caps-lock to M-x under GNU/Linux
;; From http://emacs-fu.blogspot.se/2008/12/remapping-caps-lock.html
(if (eq system-type 'gnu/linux)
    ;(shell-command "xmodmap -e 'clear Lock' -e 'keycode 66 = F13'")
  ;; Requires the bash command "xmodmap -e 'clear Lock' -e 'keycode 66 = F13'" to be run prior to this binding
  (global-set-key [f13] 'god-mode-all)
)

;; Bind Caps-lock to M-x For Windows
(if (eq system-type 'windows-nt)
    (setq w32-enable-caps-lock nil)
    ;(global-set-key [capslock] 'god-local-mode)
    (global-set-key [capslock] 'god-mode-all)
)

;;; Calender stuff

;; Add week number to Emacs calender
(copy-face font-lock-constant-face 'calendar-iso-week-face)
(set-face-attribute 'calendar-iso-week-face nil
                    :height 0.7)
(setq calendar-intermonth-text
      '(propertize
        (format "%2d"
                (car
                 (calendar-iso-from-absolute
                  (calendar-absolute-from-gregorian (list month day year)))))
        'font-lock-face 'calendar-iso-week-face))

;; End of calender stuff

;; Insert date at current position
(defun jx/current-date () (interactive)
    (insert (shell-command-to-string "echo -n $(date +%Y-%m-%d)")))

;; Insert time at current position
(defun jx/current-time () (interactive)
 (insert (shell-command-to-string "echo -n $(date +%H:%M)")))

;; Show clock in status-bar
(setq display-time t
      display-time-24hr-format t)
(display-time)

(which-key-mode)

;; Use this to automatically google thing at point
;; Shortcut is C-c / g
(require 'google-this)
(google-this-mode 1)

;; Use this to automatically translate a word or a phrase
;; Shortcut is C-ct
(require 'google-translate)
(require 'google-translate-smooth-ui)
(global-set-key "\C-ct" 'google-translate-smooth-translate)

;; Enable multiple-curors....CRAZY STUFF... http://emacsrocks.com/e13.html
(require 'multiple-cursors)
(global-set-key (kbd "C-c m c") 'mc/edit-lines)
(global-set-key (kbd "C-S-c C-S-c") 'mc/edit-lines)
(global-set-key (kbd "C->") 'mc/mark-next-like-this)
(global-set-key (kbd "C-<") 'mc/mark-previous-like-this)
(global-set-key (kbd "C-c C-<") 'mc/mark-all-like-this)
;; End multiplce cursor

;; Avoid weaselwords in bread-text when writing thesis and other articles
(require 'writegood-mode)
(global-set-key "\C-cg" 'writegood-mode)

(require 'muse-mode)     ; load authoring mode
(require 'muse-html)     ; load publishing styles
(require 'muse-latex)
(require 'muse-texinfo)
(require 'muse-docbook)
(require 'muse-project)  ; publish files in projects

;; Muse project named website, publishes to folder public_html
(setq muse-project-alist
      '(("website" ("~/Pages" :default "index")
         (:base "html" :path "~/public_html")
         (:base "pdf" :path "~/public_html/pdf"))))

;;; Super-smart Capitalization
;;;
;;; Experimental (need to test it out)
;;;
;; From http://endlessparentheses.com/super-smart-capitalization.html
;; Capitalize word, take in respect sentence construction
(defun endless/convert-punctuation (rg rp)
  "Look for regexp RG around point, and replace with RP.
Only applies to text-mode."
  (let ((f "\\(%s\\)\\(%s\\)")
        (space "?:[[:blank:]\n\r]*"))
    ;; We obviously don't want to do this in prog-mode.
    (if (and (derived-mode-p 'text-mode)
             (or (looking-at (format f space rg))
                 (looking-back (format f rg space))))
        (replace-match rp nil nil nil 1))))

(defun endless/capitalize ()
  "Capitalize region or word.
Also converts commas to full stops, and kills
extraneous space at beginning of line."
  (interactive)
  (endless/convert-punctuation "," ".")
  (if (use-region-p)
      (call-interactively 'capitalize-region)
    ;; A single space at the start of a line:
    (when (looking-at "^\\s-\\b")
      ;; get rid of it!
      (delete-char 1))
    (call-interactively 'subword-capitalize)))

(defun endless/downcase ()
  "Downcase region or word.
Also converts full stops to commas."
  (interactive)
  (endless/convert-punctuation "\\." ",")
  (if (use-region-p)
      (call-interactively 'downcase-region)
    (call-interactively 'subword-downcase)))

(defun endless/upcase ()
  "Upcase region or word."
  (interactive)
  (if (use-region-p)
      (call-interactively 'upcase-region)
    (call-interactively 'subword-upcase)))

(global-set-key "\M-c" 'endless/capitalize)
(global-set-key "\M-l" 'endless/downcase)
(global-set-key "\M-u" 'endless/upcase)

(global-set-key "\C-t" #'transpose-lines)
(define-key ctl-x-map "\C-t" #'transpose-chars)

(projectile-global-mode)

; Add this to your init file and flx match will be enabled for ido.

(require 'flx-ido)
(ido-mode 1)
(ido-everywhere 1)
(flx-ido-mode 1)
;; disable ido faces to see flx highlights.
(setq ido-enable-flex-matching t)
(setq ido-use-faces nil)

;; START Compilation support
;; Let Emacs guess the compilation argument
(require 'smart-compile)
(global-set-key "\C-cc" 'smart-compile)

;; Recompile on save
(require 'recompile-on-save)
(recompile-on-save-advice compile)

;; Auto-update tags file
(autoload 'turn-on-ctags-auto-update-mode "ctags-update" "turn on `ctags-auto-update-mode'." t)
(add-hook 'c-mode-common-hook  'turn-on-ctags-auto-update-mode)

(require 'ibuffer-git)
;; Customization for ibuffer-git is installed in custom.el

;; (require 'ibuffer-projectile)
;; (add-hook 'ibuffer-hook
;;     (lambda ()
;;       (ibuffer-projectile-set-filter-groups)
;;       (unless (eq ibuffer-sorting-mode 'alphabetic)
;;         (ibuffer-do-sort-by-alphabetic))))

;; Better defaults for occur
(defun occur-dwim ()
  "Call `occur' with a sane default."
  (interactive)
  (push (if (region-active-p)
            (buffer-substring-no-properties
             (region-beginning)
             (region-end))
          (thing-at-point 'symbol))
        regexp-history)
  (call-interactively 'occur))
(global-set-key (kbd "M-s o") 'occur-dwim)

(add-hook 'prog-mode-hook 'flyspell-prog-mode)

;; --- START ediff
;; Customize ediff to be usable
;; Got this tips from
;; www.oremacs.com/2015/01/17/setting-up-ediff
;;
(defmacro csetq (variable value)
  `(funcall (or (get ',variable 'custom-set)
                'set-default)
            ',variable ,value))

;; Setup frames the correct way
(csetq ediff-window-setup-function 'ediff-setup-windows-plain)

; Ignore whitespace
(csetq ediff-diff-options "-w")

;; Setup window configuration
(csetq ediff-split-window-function 'split-window-horizontally)

;; Restoring windows after quitting ediff
(add-hook 'ediff-after-quit-hook-internal 'winner-undo)

;; Changing keyindings
(defun ora-ediff-hook()
  (ediff-setup-keymap)
  (define-key ediff-mode-map "j" 'ediff-next-difference)
  (define-key ediff-mode-map "k" 'ediff-previous-difference))
(add-hook 'ediff-mode-hook 'ora-ediff-hook)

;; --- End of ediff-configuration

;;; --- START C++
(require 'auto-complete-clang-async)

(defun ac-cc-mode-setup ()
  (setq ac-clang-complete-executable "~/.emacs.d/personal/modules/clang-complete")
  (setq ac-sources '(ac-source-clang-async))
  (ac-clang-launch-completion-process)
  )
(defun ac-common-setup ()
  ())
(defun my-ac-config ()
  (add-hook 'c-mode-common-hook 'ac-cc-mode-setup)
  (add-hook 'auto-complete-mode-hook 'ac-common-setup)
  (global-auto-complete-mode t))

(my-ac-config)

;; (require 'auto-complete-clang-async)

;; (defun ac-cc-mode-setup ()
;;   (setq ac-clang-complete-executable "~/.emacs.d/personal/modules/clang-complete")
;;   (setq ac-sources '(ac-source-clang-async))
;;   (ac-clang-launch-completion-process)
;;   )

;; (defun my-ac-config ()
;;   (add-hook 'c-mode-common-hook 'ac-cc-mode-setup)
;;   (add-hook 'auto-complete-mode-hook 'ac-common-setup)
;;   (global-auto-complete-mode t))

;; (my-ac-config)


;; CMake support
(require 'cmake-mode)
;; More advanced syntax highlighting
(autoload 'cmake-font-lock-activate "cmake-font-lock" nil t)
(add-hook 'cmake-mode-hook 'cmake-font-lock-activate)

;; Ease use of out-of-tree build in CMake
(require 'cmake-project)
(defun maybe-cmake-project-hook ()
  (if (file-exists-p "CMakeLists.txt") (cmake-project-mode)))
(add-hook 'c-mode-hook 'maybe-cmake-project-hook)
(add-hook 'c++-mode-hook 'maybe-cmake-project-hook)

;; Toggle between implementation and test file
(require 'toggle-test)

(add-to-list 'tgt-projects '((:root-dir "~/Projects/TestDriven/MySoundex")
                             (:src-dirs "src")
                             (:test-dirs "test")
                             (:test-prefixes "Test")))

(global-set-key (kbd "C-c x t") 'tgt-toggle)
; (setq tgt-open-in-new-window <'nil or t>)

;; Toggle between implentation and header
; # Disabled because not in Melpa
; (require 'toggle-header-impl)
; (global-set-key (kbd "C-c x h") 'djw-c-toggle-impl-header-view)

; # Disabled because not in Melpa
; (require 'smarter-compile)
; (defun jonix/bindkey-compile ()
;  "Bind C-c C-c to `compile'."
;  (local-set-key (kbd "C-c C-c") 'smarter-compile))
;(add-hook 'c-mode-common-hook 'jonix/bindkey-compile)
; (add-hook 'c++-mode-hook 'jonix/bindkey-compile)

;(eval-after-load 'C++-mode
;   (define-key c++-mode-map (kbd "C-c C-c") 'smarter-compile))
;(eval-after-load 'c
;(define-key c++-mode-map (kbd "C-c C-c") 'smarter-compile)

;; Code browsing using ECB
(require 'ecb)
                                        ;(require 'ecb-autoloads)
(setq ecb-layout-name "left15")
(setq ecb-show-sources-in-directories-buffer 'always)

;;; --- END C++

;;; --- START PYTHON
;(add-hook 'python-mode-hook 'jedi:setup)
(setq jedi:setup-keys t)
(setq jedi:complete-on-dot t)
(elpy-enable)

(defun jx/execute-python ()
  (interactive)
  (python-shell-send-buffer)
  (python-shell-switch-to-shell)
  )
(eval-after-load "python"
  '(progn
     (define-key python-mode-map (kbd "<f5>") 'jx/execute-python)
     (define-key python-mode-map (kbd "C-h f") 'python-eldoc-at-point)
     ))

;; To enforce pyp8 style-rules, rewrite buffers
(require 'py-autopep8)
(add-hook 'elpy-mode-hook 'py-autopep8-enable-on-save)


;;; --- END PYTHON

;;; --- START CUCUMBER


(require 'feature-mode)
(add-to-list 'auto-mode-alist '("\.feature$" . feature-mode))
(require 'cucumber-goto-step)
(setq feature-use-rvm t)
;;; --- END CUCUMBER

;; Enable simple lookup for programming searches
(require 'howdoi)

;; Ability to swap places of buffers
;; Bound to Ctrl-Windows-<arrow>
;; Note that Windows key is little s
(require 'buffer-move)
(global-set-key (kbd "<C-s-up>")     'buf-move-up)
(global-set-key (kbd "<C-s-down>")   'buf-move-down)
(global-set-key (kbd "<C-s-left>")   'buf-move-left)
(global-set-key (kbd "<C-s-right>")  'buf-move-right)

;; Window switching. (C-x o goes to the next window)
(global-set-key (kbd "C-x O") (lambda ()
                                (interactive)
                                (other-window -1))) ;; back one

;; Make Split window show two different buffers
;; Copied from http://www.reddit.com/r/emacs/comments/25v0eo/you_emacs_tips_and_tricks/chldury
(defun vsplit-last-buffer (prefix)
  "Split the window vertically and display the previous buffer."
  (interactive "p")
  (split-window-vertically)
  (other-window 1 nil)
  (if (= prefix 1)
    (switch-to-next-buffer)))
(defun hsplit-last-buffer (prefix)
  "Split the window horizontally and display the previous buffer."
  (interactive "p")
  (split-window-horizontally)
  (other-window 1 nil)
  (if (= prefix 1) (switch-to-next-buffer)))
(global-set-key (kbd "C-x 2") 'vsplit-last-buffer)
(global-set-key (kbd "C-x 3") 'hsplit-last-buffer)

;; START ace
;; Enable very handy jump within a buffer using  Ctrl-c Space
(autoload
  'ace-jump-mode
  "ace-jump-mode"
  "Emacs quick move minor mode"
  t)
;; you can select the key you prefer to
(define-key global-map (kbd "C-c C-j") 'ace-jump-mode)

;
;; enable a more powerful jump back function from ace jump mode
(autoload
  'ace-jump-mode-pop-mark
  "ace-jump-mode"
  "Ace jump back:-)"
  t)
(eval-after-load "ace-jump-mode"
  '(ace-jump-mode-enable-mark-sync))
(define-key global-map (kbd "C-x SPC") 'ace-jump-mode-pop-mark)

(custom-set-faces
 '(aw-leading-char-face
   ((t (:inherit ace-jump-face-foreground :height 3.0)))))

;; END ace

; (require 's3c)

;; Yasnippet
(require 'yasnippet)
(yas-global-mode 1)

;; Personal Yasnippet directory outside of Melpa
;; Default is disabled
(setq yas-snippet-dirs '("~/.emacs.d/personal/yasnippet"))

;; Interact with tmux terminal emulator
(require 'emamux)
